### PR TITLE
QUAD-108 Created project statistics page and displayed d3 bar graph w…

### PR DESCRIPTION
**Pull request by bharathsharma**
QUAD-108 Created project statistics page and displayed d3 bar graph with dummy data

### CODE REVIEW

**Comment by jdamerow** [comment]
please resolve conflicts

**Comment by bharathsharma** [comment]
Merged  and conflicts resolved .

**Comment by jdamerow** [comment]
Please resolve conflicts and remove .DS_Store file.

**Comment by jdamerow** [comment]
Please remove the .DS_Store file

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java [hunk 1]
```diff
@@ -216,4 +216,34 @@ public String visualizeAllNetworks(@PathVariable("projectUnixName") String proje
...
+	 * @author Bharath Srikantan
+	 * @param projectUnixName	The project unix name
+	 * @param model				Model
+	 * @return view
+	 * @throws JAXBException
+	 * @throws QuadrigaStorageException
+	 * @throws JSONException 
+	 */
+	@RequestMapping(value = "sites/{projectUnixName}/statistics", method = RequestMethod.GET)
+	public String showProjectStatistics(@PathVariable("projectUnixName") String projectUnixName,
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java]: This method should be in its own controller.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java [hunk 2]
```diff
@@ -216,4 +216,34 @@ public String visualizeAllNetworks(@PathVariable("projectUnixName") String proje
...
+									   Model model)throws JAXBException, QuadrigaStorageException, JSONException {
+		IProject project = getProjectDetails(projectUnixName);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java]: This is just one line of code, don't use an extra method for that.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java [hunk 3]
```diff
@@ -216,4 +216,34 @@ public String visualizeAllNetworks(@PathVariable("projectUnixName") String proje
...
+
+		if (project == null) {
+			return "auth/accessissue";
+		}
+ 
+		String data = "[{\"conceptId\":\"ID-1\",\"description\":\"Desc1\",\"frequency\":\".01\"},{\"conceptId\":\"ID-2\",\"description\":\"Desc2\",\"frequency\":\".03\"},{\"conceptId\":\"ID-3\",\"description\":\"Desc3\",\"frequency\":\".01\"},{\"conceptId\":\"ID-4\",\"description\":\"Desc4\",\"frequency\":\".02\"},{\"conceptId\":\"ID-5\",\"description\":\"Desc5\",\"frequency\":\".04\"}]";
+		System.out.println("Json obj "+data);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java]: remove

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java [hunk 4]
```diff
@@ -216,4 +216,34 @@ public String visualizeAllNetworks(@PathVariable("projectUnixName") String proje
...
+	@RequestMapping(value = "sites/{projectUnixName}/statistics", method = RequestMethod.GET)
+	public String showProjectStatistics(@PathVariable("projectUnixName") String projectUnixName,
+									   Model model)throws JAXBException, QuadrigaStorageException, JSONException {
+		IProject project = getProjectDetails(projectUnixName);
+
+		if (project == null) {
+			return "auth/accessissue";
+		}
+ 
+		String data = "[{\"conceptId\":\"ID-1\",\"description\":\"Desc1\",\"frequency\":\".01\"},{\"conceptId\":\"ID-2\",\"description\":\"Desc2\",\"frequency\":\".03\"},{\"conceptId\":\"ID-3\",\"description\":\"Desc3\",\"frequency\":\".01\"},{\"conceptId\":\"ID-4\",\"description\":\"Desc4\",\"frequency\":\".02\"},{\"conceptId\":\"ID-5\",\"description\":\"Desc5\",\"frequency\":\".04\"}]";
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java]: use a json library to do that.

Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/WebsiteProjectController.java]: never mind, disregards last comment.

File: Quadriga/src/main/webapp/WEB-INF/views/projectstatistics.jsp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+{
+	d3projectstatistics(data);
+}
+</script>
+</head>
+<body>
+	<h2>Project Statistics</h2>
+	<div id='stats'></div>
+</body>
+</html>
```
Comment (jdamerow) [inline Quadriga/src/main/webapp/WEB-INF/views/projectstatistics.jsp]: don't provide a complete HTML page (with html, body, ... tags). The skeleton does that.

File: Quadriga/src/main/webapp/WEB-INF/views/public/navbar.jsp [hunk 6]
```diff
@@ -1,10 +1,6 @@
...
-<li ${currentPage == "networksearch" ? "class=\"active\"" : ""}><a
-        href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/search">Search</a></li>
-<li ${currentPage == "getprojectstatistics" ? "class=\"active\"" : ""}><a
-        href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/statistics">Statistics</a></li>
\ No newline at end of file
+<li ${currentPage == "home" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}">Home</a></li>
+<li ${currentPage == "networks" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/browsenetworks">Networks</a></li>
+<li ${currentPage == "exploregraph" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/networks">Explore</a></li>
+<li ${currentPage == "exploregraph" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/statistics">Statistics</a></li>
+<li ${currentPage == "networksearch" ? "class=\"active\"" : ""}><a href="">Search</a></li>
```
Comment (jdamerow) [inline Quadriga/src/main/webapp/WEB-INF/views/public/navbar.jsp]: what about the search url?

File: Quadriga/src/main/webapp/resources/txt-layout/js/d3projectstatistics.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
+/**
+ * @Author : Bharath Srikantan
+ * 
+ * Data visualization in D3 JQuery.
+ * This javascript should work on displaying concept collections on statistic page
+ * 
+ */
+
+function d3projectstatistics(data) {
```
Comment (jdamerow) [inline Quadriga/src/main/webapp/resources/txt-layout/js/d3projectstatistics.js]: Let's use proper camel case here.

Comment (jdamerow) [inline Quadriga/src/main/webapp/resources/txt-layout/js/d3projectstatistics.js]: and indentation

**Comment by jdamerow** [comment]
See comments. Please reopen PR when you're done.

**Comment by bharathsharma** [comment]
All issues addressed.

**Comment by jdamerow** [comment]
Cool. Please resolve conflicts and connect it with the backend.

**Comment by bharathsharma** [comment]
Merged with story/QUAD-114 and resolved conflicts
@ajaymodi FYR

**Comment by jdamerow** [comment]
Still: please remove .DS_Store file.

File: Quadriga/src/main/webapp/WEB-INF/views/public/navbar.jsp [hunk 8]
```diff
@@ -1,10 +1,6 @@
...
-        href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/networks">Explore</a></li>
-<li ${currentPage == "networksearch" ? "class=\"active\"" : ""}><a
-        href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/search">Search</a></li>
-<li ${currentPage == "getprojectstatistics" ? "class=\"active\"" : ""}><a
-        href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/statistics">Statistics</a></li>
\ No newline at end of file
+<li ${currentPage == "home" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}">Home</a></li>
+<li ${currentPage == "networks" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/browsenetworks">Networks</a></li>
+<li ${currentPage == "exploregraph" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/networks">Explore</a></li>
+<li ${currentPage == "exploregraph" ? "class=\"active\"" : ""}><a href="${pageContext.servletContext.contextPath}/sites/${project.unixName}/statistics">Statistics</a></li>
```
Comment (jdamerow) [inline Quadriga/src/main/webapp/WEB-INF/views/public/navbar.jsp]: you want the Statistics menu entry to be active when you are on the statistics page, so you should use your own currentPage value.

**Comment by jdamerow** [comment]
ok to test

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/ProjectStatsController.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Autowired
+    private ProjectStats projectStats;
+
+    @Autowired
+    private Environment env;
+
+    private int getCount(List<IConceptStats> conceptList) {
+
+        int cnt = Integer.parseInt(env.getProperty("project.stats.topcount"));
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/ProjectStatsController.java:51]: This line raises the following exception:

```
edu.asu.spring.quadriga.exceptions.QuadrigaException: null
    java.lang.Integer.parseInt(Integer.java:542)
    java.lang.Integer.parseInt(Integer.java:615)
    edu.asu.spring.quadriga.web.publicwebsite.ProjectStatsController.getCount(ProjectStatsController.java:51)
```

**Comment by jdamerow** [comment]
please reopen PR when you're done

**Comment by bharathsharma** [comment]
Resolved all issues

**Comment by jdamerow** [comment]
retest this please

**Comment by jdamerow** [comment]
Great start! A few things and its good to go:
- You need to use the skeleton for public pages (with the blue navigation bar).
- The diagram should be centered.
- the layout of the hover over pop up should be inline with the rest of the layout of the page (no red font, maybe a little smaller font), and the pop ups should be smaller and more squared. Right now it also runs out of the page (see screenshot).

image

**Comment by bharathsharma** [comment]
Addressed all issues.Attached screen shot for reference.

screen shot 2016-03-31 at 10 40 20 pm

**Comment by jdamerow** [comment]
ok to test

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/impl/ProjectStats.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * This class represents business logic for getting top concepts from the given
+ * list of networks.
+ *
+ * @author ajaymodi
+ *
+ */
+
+@Service
+public class ProjectStats implements IProjectStats {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/impl/ProjectStats.java:31]: @ajaymodi, please write a test class for this class.

**Comment by jdamerow** [comment]
Looks great. Is there a reason though, that the diagram is still not centered?

**Comment by bharathsharma** [comment]
I was centering the diagram with respect to first bar. Ill fix it in next release

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/impl/ProjectStats.java (hunk 10 above)
Comment (ajaymodi) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/impl/ProjectStats.java:31]: I have added test class.

File: Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/ProjectStatsTest.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        ConceptStats cs1 = new ConceptStats("url1", null, null, 3);
+        ConceptStats cs2 = new ConceptStats("url2", null, null, 1);
+        List<IConceptStats> conceptStatsList = new ArrayList<IConceptStats>();
+        conceptStatsList.add(cs1);
+        conceptStatsList.add(cs2);
+        
+        Mockito.when(mockedTransformationManager.getTransformedNetwork("id")).thenReturn(transformedNetwork1);
+        Mockito.when(mockedTransformationManager.getTransformedNetwork("id2")).thenReturn(transformedNetwork2);
+                
+        assertEquals("conceptCount", conceptStatsList, projectStatsUnderTest.getConceptCount(networkList));
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/ProjectStatsTest.java:84]: YOu should also test if the sorting works correctly. Write a separate test method for that and it probably makes sense to put the set up code into a before method.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/ProjectStatsController.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+     * @param model
+     *            Model
+     * @return view
+     * @throws JAXBException
+     * @throws QuadrigaStorageException
+     */
+    @RequestMapping(value = "sites/{projectUnixName}/statistics", method = RequestMethod.GET)
+    public String showProjectStatistics(
+            @PathVariable("projectUnixName") String projectUnixName, Model model)
+            throws JAXBException, QuadrigaStorageException {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/ProjectStatsController.java:88]: There are conflicts in your code. Please merge them. Also, there is now an aspect that injects the correct project given the unix name. Try to incorporate that:
https://diging.atlassian.net/wiki/display/QUAD/Injecting+projects+by+unix+names

You can then loose the next line.

**Comment by jdamerow** [comment]
please resolve conflicts and incorporate comments. Reopen PR when you're done.

**Pull request closed without merging**